src/artemis/devices/fast_grid_scan.py

```diff
@@ -0,0 +1,144 @@
+            time_remaining = None
+        else:
+            time_remaining = time_elapsed / fraction
+        for watcher in self._watchers:
+            watcher(
+                name=self._name,
+                current=value,
+                initial=0,
+                target=self._target_count,
+                unit="images",
+                precision=0,
+                fraction=fraction,
+                time_elapsed=time_elapsed,
+                time_remaining=time_remaining,
+            )
+
+    def _running_changed(self, value=None, old_value=None, **kwargs):
+        if (old_value == 1) and (value == 0):
+            # Stopped running
+            number_of_images = self.device.position_counter.get()
+            if number_of_images != self._target_count:
+                self.set_exception(
+                    Exception(
+                        f"Grid scan finished without collecting expected number of images. Expected {self._target_count} got {number_of_images}."
+                    )
+                )
+            else:
+                self.set_finished()
+            self.clean_up()
+
+    def clean_up(self):
+        self.device.position_counter.clear_sub(self._notify_watchers)
+        self.device.status.clear_sub(self._running_changed)
+
+
+class FastGridScan(Device):
+
+    x_steps: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "X_NUM_STEPS")
+    y_steps: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y_NUM_STEPS")
+    z_steps: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z_NUM_STEPS")
+
+    x_step_size: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "X_STEP_SIZE")
+    y_step_size: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y_STEP_SIZE")
+    z_step_size: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z_STEP_SIZE")
+
+    dwell_time: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "DWELL_TIME")
+
+    x_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "X_START")
+    y1_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y_START")
+    y2_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Y2_START")
+    z1_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z_START")
+    z2_start: EpicsSignalWithRBV = Component(EpicsSignalWithRBV, "Z2_START")
+
+    position_counter: EpicsSignal = Component(
+        EpicsSignal, "POS_COUNTER", write_pv="POS_COUNTER_WRITE"
+    )
+    x_counter: EpicsSignalRO = Component(EpicsSignalRO, "X_COUNTER")
+    y_counter: EpicsSignalRO = Component(EpicsSignalRO, "Y_COUNTER")
+    scan_invalid: EpicsSignalRO = Component(EpicsSignalRO, "SCAN_INVALID")
+
+    run_cmd: EpicsSignal = Component(EpicsSignal, "RUN.PROC")
+    stop_cmd: EpicsSignal = Component(EpicsSignal, "STOP.PROC")
+    status: EpicsSignalRO = Component(EpicsSignalRO, "SCAN_STATUS")
+
+    # Kickoff timeout in seconds
+    KICKOFF_TIMEOUT: float = 5.0
+
+    def set_program_data(self, nx, ny, width, height, exptime, startx, starty, startz):
```

Comment:
Kinda feel like more structured API here would help? `set_program_data` call in the test just looks like pushing a bunch of random numbers

Reply:
Yh, I left this in as a bit of a placeholder until I had a better feel for how we would actually be using the Device upstream but I'll think about improving it now.

Reply:
I borrowed some of the work of @callumforrester for this, which also included some checking against motor limits.

Reply:
Didn't spot this chain, I think this needs updating to reflect this thread: https://github.com/DiamondLightSource/python-artemis/pull/2#discussion_r780119290